### DISCUSSION TITLE
Proxmox Setup

### DISCUSSION THREAD

**Discussion by Bojack84** [aaronwmorris/indi-allsky#932] [Q&A]
I am trying to set this up in proxmox and running into some issues.  I have went through the install process and when I get to the ./setup.sh I get an error.  Has anyone set this up in Proxmox and if so can you please help me?  Thanks!

**Comment by aaronwmorris** [comment]
Are you setting up in a VM or a container?  What error are you seeing?

**Comment by Bojack84** [comment]
I am setting it up as a Debian LXC container and I get this error below. 
![image](https://github.com/aaronwmorris/indi-allsky/assets/23743816/ce1c2cb4-1f36-407f-a332-19a6d8a608fe)

**Reply by aaronwmorris** [reply]
These packages are usually already available, but in rare cases like this you just need to manually install.

```
sudo apt-get install lsb-release libc-bin whiptail
```

The real challenge for you will be compiling INDI.  There are no up to date packages available for INDI for Debian.  You can manually compile with `./misc/build_indi.sh`.  If you are willing to switch to Ubuntu, you can use the INDI PPA (which is automatically used in setup.sh).

**Reply by Bojack84** [reply]
Yeah it would not be any problem to switch to Ubuntu.  Which way would be better to operate on.

**Comment by Bojack84** [comment]
So I deleted and started fresh with Ubuntu in LXC and now I am getting the following error: 
![image](https://github.com/aaronwmorris/indi-allsky/assets/23743816/cd5ad2e4-a19b-499e-86ea-3867a9f4d6ce)
![image](https://github.com/aaronwmorris/indi-allsky/assets/23743816/39d03488-1740-405d-8373-38286301277c)

**Reply by Bojack84** [reply]
Just an FYI I am trying to set this up to use it with an IP camera.

**Reply by aaronwmorris** [reply]
Install this package:

```
sudo apt-get install software-properties-common
```

**Reply by aaronwmorris** [reply]
What kind of IP camera?

When you get to the indi server setup, make sure to select `indi_webcam_ccd`

**Reply by Bojack84** [reply]
Its a reolink camera.

**Reply by aaronwmorris** [reply]
Great!  Hopefully you should be able to directly use my indi config example.

**Reply by Bojack84** [reply]
So I get here 

![image](https://github.com/aaronwmorris/indi-allsky/assets/23743816/e4af7b3f-81b6-4be3-8413-d4528209d740)

I pick number 1 and then I get this at the end. 

![image](https://github.com/aaronwmorris/indi-allsky/assets/23743816/625235a3-8300-4b6d-9e14-7cdefeb238a5)

**Reply by aaronwmorris** [reply]
You just need to reboot and re-run setup.sh.  This is a normal error condition on the first run.

**Reply by Bojack84** [reply]
So I do that and at the end it comes up and says it is already installed would I like to upgrade the software.  I say no and then I get the same dbus user error.

**Reply by aaronwmorris** [reply]
This is one of the troubles with container environments.  None of the normal system management tools that are commonly available operate correctly.

Container environments normally expect one process per container.  If you take a look at the docker/ folder, you can see it requires quite a few containers to run everything.

If you cannot use docker, lxc might not be the best way to move forward at this time.

**Reply by Bojack84** [reply]
Ok I can try that method as well.  I have a way to spin up a docker env.  I will let you know how that goes.  Thanks for all your assistance on this.  I appreciate it.

**Reply by aaronwmorris** [reply]
I am definitely interested in supporting other container environments like lxc, I am just not sure lxc has the necessary functionality to enable a complex automated deployment (like dockerfiles and docker compose).

**Comment by Bojack84** [comment]
Ok so I am back.  I switched to a raspberry pi.  I am trying to install everything but I am lost on where to put the code for the ipcamera? I am assuming that I need to update the file before I select a camera interface?  Thanks for you assistance in advance.

**Reply by aaronwmorris** [reply]
You are in luck.  I have implemented a new camera interface called the `pyCurl Camera` and you can put the URL directly in the configuration for that camera (near the bottom on the Camera tab).  This avoids the more complex indi_webcam_ccd configuration.

**Reply by Bojack84** [reply]
Ok so if I am understanding you correctly I can pick number 4 in the listings?

**Reply by aaronwmorris** [reply]
That is correct.

**Reply by Bojack84** [reply]
Thank you.  I am running through everything now.  Going to try it and let you know.

**Reply by Bojack84** [reply]
I got it up and running on a raspberryPI.  Now all I need to do is get a good poe camera and do it like this here:  https://github.com/aaronwmorris/indi-allsky/discussions/949  
I appreciate all you help on this.

**Reply by Bojack84** [reply]
Where can I update the resolution?

**Reply by aaronwmorris** [reply]
With an IP webcam, you have to adjust the resolution on the camera itself.  There are no external controls to adjust the camera from indi-allsky.

**Reply by Bojack84** [reply]
Thank you. I'm getting errors now and it's not capturing images. I think from reading through the discussions that it's an issue with the sqlite being corrupted.

**Reply by Bojack84** [reply]
Right after I said this I was reading and I realized that I set the exposure period day/night below the max after reading the little descriptions. I assume that was making the memory just bog down.

**Comment by Bojack84** [comment]
So I have had the raspberrypi running for a while and it worked great till the SD card died and I had to reload it and I set all the settings back to  what I think they were before and now it crashes every time it does the star trails/video creation.  Below is a screenshot of what it is set to currently.  
![image](https://github.com/aaronwmorris/indi-allsky/assets/23743816/e8b7cf41-c774-4c91-8b81-12a53722e1a9)

**Reply by aaronwmorris** [reply]
Is it running out of memory?  Do you know what error is occurring?

**Reply by Bojack84** [reply]
The Log does not go back further enough.  I guess I will have to check it in the morning after it fails. Unless  you can tell me another way.

**Reply by Bojack84** [reply]
So a heads up.  It didnt fail this morning, go figure, so I tried to do ubuntu in Proxmox as a VM and it is up and running.  I am going to keep the raspberry pi running as well as the VM. I will keep you updated as how that goes.  So far it seams to be a lot smoother than the raspberry pi.

**Reply by aaronwmorris** [reply]
If you are still using the Pi3, the VM will certainly have much higher performance.  Using the Pi3 as a remote indiserver for the VM could be a much better architecture.

**Reply by Bojack84** [reply]
I am not sure what you mean by that but would be willing to learn. Do you have a link I can read up on?

**Reply by aaronwmorris** [reply]
INDI is a full client/server implementation.  Meaning INDI clients (like indi-allsky) do not need to run on the same hardware where the camera is connected.  Only the indiserver needs to be running on the same hardware.

In the indi-allsky config, on the Camera tab, you can change the indiserver hostname to your raspberry pi and just download the images remotely from that indiserver, but have all of the speed and CPU benefits of the VM.

This is a rough diagram:
![indi-allsky_modes-Remote](https://github.com/aaronwmorris/indi-allsky/assets/17464290/20d6bc07-943e-46ef-8802-64567766c51f)

**Reply by Bojack84** [reply]
Ok that is interesting.  So I can have the raspberrypi do the colleting of the images and the VM do the processing? I am going to research that.

**Reply by wjschmitt** [reply]
Hmm, interesting.  So if I spin up a proxmox ubuntu virtual machine, I can run setup.sh on both the vm and on a RPi that has the camera attached, correct?  Then I just run indi-allsky on the vm and indiserver on the Pi, right?  Or does indi-allsky and indiserver run on both?  It seems like both services have to run on both the vm and the pi.  Do I configure the file transfer tab on the pi to point to the vm, and configure the camera tab on the vm to point to the pi?   What about the file transfer tab?  Are all image files stored on the vm?  Sorry, enough questions.  I'll just give it a go and see what happens.

**Reply by aaronwmorris** [reply]
Only the indiserver needs to run on the Raspi.  On the VM, only the indi-server service needs to run.  You just point the indi-allsky service at the indiserver in the Camera Tab (save and restart).

All of the images will be stored on the VM.  You do not need to configure any file transfer settings for this to work.

**Reply by wjschmitt** [reply]
Ok, I think I have this working, although what you said is confusing.  Indiserver runs on the Pi, indi-allsky service runs on the vm, right?
OK, I do that and I can now display the webpage when pointing to the vm but no images are being displayed yet.  How can I check that the indiserver is sending images to the vm?  I see a lot of Capture Worker errors in the log.  I checked that port 7624 is open on both ends.

**Reply by aaronwmorris** [reply]
If a new camera was defined, I would make sure you have the correct camera selected on the left hand side drop down.

**Reply by wjschmitt** [reply]
It's the same camera that I used when indi-allsky service was on the Pi.  When I look through the logs, I see errors about the capture worker and that libcamera-still cannot be found.  I'm thinking that libcamera has not been installed on my vm running ubuntu.

**Reply by aaronwmorris** [reply]
What camera are you using?

**Comment by wjschmitt** [comment]
That's great.  I'll give it a go.  Thanks.Sent from my Galaxy
-------- Original message --------From: Aaron W Morris ***@***.***> Date: 3/8/24  8:37 PM  (GMT-05:00) To: aaronwmorris/indi-allsky ***@***.***> Cc: wjschmitt ***@***.***>, Comment ***@***.***> Subject: Re: [aaronwmorris/indi-allsky] Proxmox Setup (Discussion #932) 
Only the indiserver needs to run on the Raspi.  On the VM, only the indi-server service needs to run.  You just point the indi-allsky service at the indiserver in the Camera Tab (save and restart).
All of the images will be stored on the VM.  You do not need to configure any file transfer settings for this to work.

**Comment by wjschmitt** [comment]
Do I need to set up the SyncAPI tab as well to make this work?

**Reply by aaronwmorris** [reply]
No, the SyncAPI is not used.  The only setting you need to change is the `INDI Server` setting on the Camera Tab (and restart).  As long as the indiserver on the Raspi is connected to the camera, it should work.

**Comment by wjschmitt** [comment]
RPi HQ. - imx477Sent from my Galaxy
-------- Original message --------From: Aaron W Morris ***@***.***> Date: 3/11/24  4:19 PM  (GMT-05:00) To: aaronwmorris/indi-allsky ***@***.***> Cc: wjschmitt ***@***.***>, Comment ***@***.***> Subject: Re: [aaronwmorris/indi-allsky] Proxmox Setup (Discussion #932) 
What camera are you using?

**Reply by aaronwmorris** [reply]
I see the problem.  The remote indiserver option only works with cameras connected via INDI, not cameras using libcamera-still directly

The indi_libcamera_ccd driver is still not quite 100% stable yet, but if you can figure out how to get it build on your Raspi, it will probably work.  The other option is [indi_pylibcamera](https://pypi.org/project/indi-pylibcamera/), but it can be a bit complex to get setup, but it should work relatively well with the HQ camera.

**Reply by wjschmitt** [reply]
I was able to build and install the indi_pylibcamera.  When I run indiserver -v indi_pylibcamera  I start receiving output where it appears that the indi-allsky on the vm is connecting.  Indi-allsky creates a new camera called V4L2_CCD.  In the log, I can see it connecting to the indiserver, creating an instance of an indi client, but then I see errors for the capture workers, and see no images.

So that's as far as I can get.  I guess I'll have to abandon this proxmox idea.  Unless you see something that I'm doing wrong.

**Reply by aaronwmorris** [reply]
I think you are close.  Is it possible you are still connecting to the local indiserver?  Are you sure you only have a single indiserver on the remote raspberry pi?

**Reply by wjschmitt** [reply]
I check again.  On the Config/CaCamera tab, what selection should I make for the Camera Interface - Indi or libcamera?

**Reply by aaronwmorris** [reply]
For indi_pylibcamera, you would select `INDI` and make sure the hostname is set for the hostname/ip address of the Raspi.

**Reply by aaronwmorris** [reply]
And by the way, you should get a new camera called `indi_pylibcamera` in your list.

**Reply by wjschmitt** [reply]
OK, I selected INDI, set the ip for the pi, and I see a new camera called pylibcamera Main, which I select.  Still no images displayed.  On the Config/Admin tab, I see a selection called Non-local Images.  Should that be turned on?

**Reply by aaronwmorris** [reply]
No, I will try to clarify the nonlocal images setting.  Do you have any config in the INDI config at the bottom?

**Reply by wjschmitt** [reply]
No INDI config.  But in the log, I do see this - 

Mar 15 09:45:15 ubserver [INFO] Image-1-6881/MainThread miscDb.addThumbnail() [981]: Adding thumbnail to DB: /var/www/html/allsky/images/ccd_ed2302c9-c1a0-4a3b-afaa-6e68a910d3a0/thumbnails/20240315/15_09/82a72ad4-34af-4f16-aab2-ab7d366740af.jpg
Mar 15 09:45:15 ubserver [INFO] Image-1-6881/MainThread miscDb.addImage() [198]: Adding image ccd_ed2302c9-c1a0-4a3b-afaa-6e68a910d3a0/20240315/day/15_09/ccd3_20240315_094515.jpg to DB

Does this mean the pylibcamera is working, but images are just not being displayed?  I cannot find any jpg at that location - /var/www/html/allsky/images/.

**Reply by wjschmitt** [reply]
Disregard my last.  I do see the image files in the right location on the vm.  They are being created but just not being displayed on the webpage.   And when I view them by other means, they are all blank - just the labels appear.

**Reply by aaronwmorris** [reply]
Do you happen to have that "Non-local Images" option enabled?  If you do, disable it and restart indi-allsky or reboot.  That is the only thing that could cause this behavior.

**Reply by wjschmitt** [reply]
No, non-local images is disabled.  I did a reboot.  Images are being sent from the pi to the vm, but they are not displayed on the webpage.  The images have all the labels but no camera output.

Could there be something missing on the vm web server.  Maybe a permissions issue.

**Reply by aaronwmorris** [reply]
Are you seeing blank images with the labels or no images at all?

**Reply by wjschmitt** [reply]
I see all the labels, which I assume are overlaid on the image.  So I guess I see the image.  It just does not have any camera output on it.

**Reply by aaronwmorris** [reply]
Excellent.  I am guessing the image is just a flat grey.

In this case, go ahead and override the `Camera Bit Depth` on the Camera tab to 16-bit and restart.  This should fix the grey image.

**Reply by wjschmitt** [reply]
OK we're almost there.   When I view the image directly on the vm, it is now showing what the camera produces with the labels.  But the webpage still shows nothing.   I restarted everything, including apache.  On the Latest page, nothing is displayed - no image and no labels.

**Reply by aaronwmorris** [reply]
Can you provide the output of the following:

```
source virtualenv/indi-allsky/bin/activiate

./config.py dump
```

**Reply by wjschmitt** [reply]
Here you go.  
[vm_dump.txt](https://github.com/aaronwmorris/indi-allsky/files/14621603/vm_dump.txt)

**Reply by wjschmitt** [reply]
OK, I should have done this sooner.  I created a new VM, loaded Ubuntu 22.04 and re-installed indi-allsky.  This time when I ran setup.sh, I selected INDI as the camera type.  It loaded all the correct drivers.  I wasn't sure which camera to select, so I went with indi-webcam.

Then I started indi-allsky service on the vm.  Connected it to the indiserver on the pi, selected the pylibcamera and set the bit depth to 16.  And lo and behold, everything worked.  I get good images now.  I think it was just a matter of not having all the right indi drivers.

Thanks for all your help, Aaron.

**Comment by wjschmitt** [comment]
One final problem to solve  -  I have an external script that gets temperature from my camera enclosure, which runs on the pi.   So now with indi-allsky running on the vm, it won't execute.  I think I will just modify the script on the pi to ssh the output file to the vm so it can be displayed.

**Reply by aaronwmorris** [reply]
If you can setup SSH keys for passwordless authentication, you can run the command remotely on the PI and store the output locally.

```
ssh username@hostname "/path/to/command.sh" > output.txt
```

**Reply by wjschmitt** [reply]
Yep.  Just did that.  Thanks.

**Comment by 4358haltern** [comment]
Ich habe das gleiche vor❗️ 
Welche Camera willst du verwenden❓
